### PR TITLE
added multiple choice feature

### PR DESCRIPTION
**Pull request by mehtadarshan1**
Also fixed a bug where students can go back to answered question and re-answer them and gain points.

### CODE REVIEW

File: public/css/admin.css [hunk 1]
```diff
@@ -13,7 +13,7 @@ body {
 
 #admin-content {
   max-height: 750px;
-  overflow-y: auto;
+  overflow-y: auto;
```
Comment (ahmedqarmout2) [inline public/css/admin.css]: remove space at the end of line

File: public/js/admin.js [hunk 2]
```diff
@@ -97,27 +97,62 @@ var displayQuestionTable = function() {
             $('#option-settings').removeClass('active');
         }
     });
-}
-
+}
```
Comment (ahmedqarmout2) [inline public/js/admin.js]: remove spaces

File: public/js/admin.js [hunk 3]
```diff
@@ -97,27 +97,62 @@ var displayQuestionTable = function() {
...
             });
-        }
+        }    
     });
+}   
+var mcAnswerCount = 4;
+var addMCAnswers = function(dom){ 
+    mcAnswerCount++;
+     var newdiv = document.createElement('div');
+     newdiv.innerHTML = "<p>Is Correct: <input type='radio' name='radbutton' value='" + mcAnswerCount + "'/><input class='form-control' type='text' name='mcans" + mcAnswerCount + "' placeholder='Enter Answer Here' required='required' style='float:left;'/> <button onclick='$(this).parent().remove();'>delete</button></p>"
```
Comment (ahmedqarmout2) [inline public/js/admin.js]: try string formatting instead of addition of pieces

File: public/js/admin.js [hunk 4]
```diff
@@ -97,27 +97,62 @@ var displayQuestionTable = function() {
...
+     $('#qAnswer > div.form-group').append(newdiv);
 }
 
-/* Display the application statistics form. */
+// replace the answer field in Question-creation.pug for specific question
+var getQuestionFormAnswer = function(form){
+    $.ajax({ 
+        type: 'GET',
+        url: form,
+        success: function(data){
```
Comment (ahmedqarmout2) [inline public/js/admin.js]: always include failure

Comment (ahmedqarmout2) [reply public/js/admin.js]: and handle it

File: public/js/admin.js [hunk 5]
```diff
@@ -97,27 +97,62 @@ var displayQuestionTable = function() {
...
+            $('#qAnswer').html(data);
+        }
+    });
+    
+}
+   
+
```
Comment (ahmedqarmout2) [inline public/js/admin.js]: remove extra spaces in your commit

File: public/js/admin.js [hunk 6]
```diff
@@ -323,17 +357,19 @@ var submitEditForm = function(id) {
...
     if ($('#qtext').summernote('isEmpty')) {
         $('#result').html('Please enter a question body in the editor.');
         return;
     }
     question['text'] = $('#qtext').summernote('code');
-
+    
+    // add the question type
+    question['type'] = $('#qType').select().val();
+    console.log(question);
```
Comment (ahmedqarmout2) [inline public/js/admin.js]: remove console.log

File: public/js/admin.js [hunk 7]
```diff
@@ -344,7 +380,6 @@ var submitQuestionForm = function() {
             } else if (data == 'success') {
                 $('#result').html('Question added to database');
                 setTimeout(displayQuestionTable, 1000);
```
Comment (ahmedqarmout2) [inline public/js/admin.js:430]: why do we need `setTimeout(displayQuestionTable, 1000)`, and why did you remove the modal hide

File: public/js/question.js [hunk 8]
```diff
@@ -1,10 +1,10 @@
-$('#answerform').submit(function(evt) {
+$('#re_answerform').submit(function(evt) {
     evt.preventDefault();
     var ans = $('#answer').val();
     $.ajax({
         type: 'POST',
-        url: '/submitanswer',
-        data: $('#answerform').serialize(),
+        url: '/submitREanswer',
```
Comment (ahmedqarmout2) [inline public/js/question.js]: why changing the route?

File: server.js [hunk 9]
```diff
@@ -437,6 +449,30 @@ app.post('/submitanswer', function(req, res) {
     });
 });
 
+/* check if the submitted MC answer is correct */
+app.post('/submitMCanswer', function(req, res) {
```
Comment (ahmedqarmout2) [inline server.js]: No need for a new route. Wait for my changes and you can use the same route

File: server.js [hunk 10]
```diff
@@ -542,7 +578,29 @@ app.post('/userupload', upload.single('usercsv'), function(req, res) {
  */
 app.post('/questionadd', function(req, res) {
     req.body.basePoints = parseInt(req.body.basePoints);
-    req.body.type = questions.QUESTION_REGULAR;
+    // correct the question Type in database:
+    switch(req.body.type){
+        case 're':
```
Comment (esteeroramy) [inline server.js]: can we have some documentation for the naming, something like this to standardize the use of these question types:

var question_types= Object.freeze({
    QUESTION_RE:     0x0, // regex
    QUESTION_MC:    0x1, //multiple choice
});

File: public/js/question.js [hunk 11]
```diff
@@ -21,3 +21,28 @@ $('#answerform').submit(function(evt) {
...
+
+$('#mc_answerform').submit(function(evt) {
+    evt.preventDefault();
+    var ans = $("input[name=answer]:checked").val();
+    $.ajax({
+        type: 'POST',
+        url: '/submitREanswer',
+        data: $('#mc_answerform').serialize(),
+        success: function(data) {
+            if (data.result == 'incorrect') {
```
Comment (esteeroramy) [inline public/js/question.js]: This looks like it's duplicated from submitting a short answer. can we combine the 2 and add a flag for the type?

File: views/mc-answer.pug [hunk 12]
```diff
@@ -0,0 +1,24 @@
+.form-group
+	p
+		| Is Correct:
```
Comment (esteeroramy) [inline views/mc-answer.pug]: can we use a loop here rather than duplicating code?

File: server.js [hunk 13]
```diff
@@ -193,6 +193,9 @@ const questionForm = pug.compileFile('views/question-creation.pug');
 const questionEdit = pug.compileFile('views/question-edit.pug');
 const statistics = pug.compileFile('views/statistics.pug');
 
+const regexForm = pug.compileFile('views/regex-answer.pug');
+const multiplechoiceFrom = pug.compileFile('views/mc-answer.pug');
```
Comment (Mohammed-JS-Khan) [inline server.js]: Make it camel case and fix the spelling (from instead of form) to reduce confusion

File: views/question-creation.pug [hunk 14]
```diff
@@ -1,35 +1,49 @@
...
-  .form-group
-    label(for='qpoints') Points
-    input.form-control#qpoints(type='number' name='basePoints'
-                               placeholder='Enter points'  min="1" required)
-  .form-group
-    label(for='qhint') Hint
-    input.form-control#qhint(type='text' name='hint'
+    select#qType.select.form-control.input-lg(name='qType' )
+      option(value='0' selected disabled) -- Select Question Type --
+      option(value='mc') Multiple Choice
```
Comment (esteeroramy) [inline views/question-creation.pug]: if we do standardize using `question_types` this would change

File: views/question-creation.pug [hunk 15]
```diff
@@ -1,35 +1,49 @@
...
-    label(for='successmsg') Follow up message
-    input.form-control#successmsg(type='text' name='successMessage'
-                                  placeholder='Enter message'
-                                  aria-describedby='success-msg-help')
-    p.form-text.text-muted#success-msg-help
-      | This is a message that will be displayed to the user after they
-      | successfully answer a question. It can be used to provide additional
-      | information on the topic.
-    button(class='btn btn-lg btn-primary btn-block' type='submit') Create
+    .form-group
```
Comment (esteeroramy) [inline views/question-creation.pug]: some extra spaces here

File: views/question.pug [hunk 16]
```diff
@@ -20,18 +20,33 @@ block content
...
-            button(class='btn btn-lg btn-primary btn-block' type='submit' disabled) Submit
-          else
-            input.form-control#answer(name='answer' placeholder='Enter your answer here'
-                                      required autofocus)
-            button(class='btn btn-lg btn-primary btn-block' type='submit') Submit
+
+        h2.qtext #{question.title}
+        h5.qtext !{question.text}
+        case question.type
+          when 0
```
Comment (esteeroramy) [inline views/question.pug]: what is 0 and what is 1, documentation please

File: server.js [hunk 17]
```diff
@@ -437,6 +449,30 @@ app.post('/submitanswer', function(req, res) {
...
+            data.points = req.session.question.basePoints;
+            /* remove question from questions list, TODO: fetch new one */
+            for (var ind in req.session.questions) {
+                if (req.session.questions[ind].id == req.session.question.id)
+                    break;
+            }
+            req.session.questions.splice(ind, 1);
+            req.session.questionAnswered = true;
+        } else {
+            /* TODO: reload sidebar stats, save in data.html */
```
Comment (esteeroramy) [inline server.js]: side question: do we have a ticket open for this?

**Comment by esteeroramy** [comment]
Can you please add some screenshots so that we can see how this looks like? thanks!

**Review by esteeroramy — CHANGES_REQUESTED** [review]
Create question:
admin1
admin2

Student view:
stduen 1

File: public/js/admin.js (hunk 7 above)
Comment (mehtadarshan1) [reply public/js/admin.js:430]: I'm not using modal anymore for creating questions, and seTimeout was there from previous revision

File: public/js/question.js (hunk 8 above)
Comment (mehtadarshan1) [reply public/js/question.js]: Need to keep each question type to it's own route, code will become spaghetti if dont do it

File: server.js (hunk 9 above)
Comment (mehtadarshan1) [reply server.js]: We do need new route for every question, but okay will wait for what you have done

File: server.js (hunk 10 above)
Comment (mehtadarshan1) [reply server.js]: The naming convention was already there in server/question.js

File: public/js/admin.js (hunk 2 above)
Comment (mehtadarshan1) [reply public/js/admin.js]: what space?

File: public/js/question.js (hunk 11 above)
Comment (mehtadarshan1) [reply public/js/question.js]: I have thought about that as  well, but the problem we run in to is scalability of adding more question types, and for that each question has a different way of collecting the answer. In this case for MC and RE questions the method of checking the answer is the same.

File: views/mc-answer.pug (hunk 12 above)
Comment (mehtadarshan1) [reply views/mc-answer.pug]: yes we can!

File: server.js (hunk 13 above)
Comment (mehtadarshan1) [reply server.js]: fixed

File: views/question.pug (hunk 16 above)
Comment (mehtadarshan1) [reply views/question.pug]: documentation is in server\questions.js line 28,29

File: server.js (hunk 17 above)
Comment (mehtadarshan1) [reply server.js]: nope

File: public/js/question.js (hunk 11 above)
Comment (ahmedqarmout2) [reply public/js/question.js]: I agree with mehtadarshan1. Backend will be doing the same thing as well. Different path for different type of question

**Comment by ahmedqarmout2** [comment]
Looks good man. Well done
